File: app/src/main/java/com/dominikgaller/alpinebooking/guide/inbound/rest/GuideExceptionHandler.java
package com.dominikgaller.alpinebooking.guide.inbound.rest;

import com.dominikgaller.alpinebooking.guide.core.domain.guidetour.exception.GuideTourNotFoundException;
import com.dominikgaller.alpinebooking.guide.core.domain.guidetour.exception.InvalidGuideTourStateException;
import com.dominikgaller.alpinebooking.guide.core.domain.guidetour.exception.TourCompletedBeforeStartException;
import com.dominikgaller.alpinebooking.guide.core.domain.guidetour.exception.TourStartTooEarlyException;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

/**
 * Translates guide domain exceptions into HTTP error responses.
 *
 * <p>Each handler returns {@code { "error": "<message>" }} with the appropriate HTTP status.
 *
 * <p>SDD: See {@code documentation/use-cases/uc05-start-tour.spec.md}, section 3.
 */
@RestControllerAdvice
public class GuideExceptionHandler {

    @ExceptionHandler(GuideTourNotFoundException.class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public ErrorResponse handleGuideTourNotFound(final GuideTourNotFoundException ex) {
        return new ErrorResponse(ex.getMessage());
    }

    @ExceptionHandler(TourCompletedBeforeStartException.class)
    @ResponseStatus(HttpStatus.CONFLICT)
    public ErrorResponse handleTourCompletedBeforeStart(
            final TourCompletedBeforeStartException ex) {
        return new ErrorResponse(ex.getMessage());
    }

    /**
     * Backstop for malformed input reaching the domain as an
     * {@code IllegalArgumentException} — chiefly {@code UUID.fromString} on the
     * {@code guideTourId} path variable in {@code StartTourDriver}. A client error, and a
     * 500 before this mapping existed.
     *
     * <p>See {@code ports/start-tour.inport.spec.md} section 7 and
     * {@code coding-style.definition.md} section 6.2.
     */
    @ExceptionHandler(IllegalArgumentException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public ErrorResponse handleIllegalArgument(final IllegalArgumentException ex) {
        return new ErrorResponse(ex.getMessage());
    }

    @ExceptionHandler(InvalidGuideTourStateException.class)
    @ResponseStatus(HttpStatus.CONFLICT)
    public ErrorResponse handleInvalidGuideTourState(final InvalidGuideTourStateException ex) {
        return new ErrorResponse(ex.getMessage());
    }

    @ExceptionHandler(TourStartTooEarlyException.class)
    @ResponseStatus(HttpStatus.CONFLICT)
    public ErrorResponse handleTourStartTooEarly(final TourStartTooEarlyException ex) {
        return new ErrorResponse(ex.getMessage());
    }

    public record ErrorResponse(String error) {
    }
}

File: src/main/kotlin/com/example/contractmanagement/guide/inbound/rest/GuideExceptionHandler.kt
package com.example.contractmanagement.guide.inbound.rest

import com.example.contractmanagement.guide.core.domain.guidetour.exception.GuideTourNotFoundException
import com.example.contractmanagement.guide.core.domain.guidetour.exception.InvalidGuideTourStateException
import com.example.contractmanagement.guide.core.domain.guidetour.exception.TourStartTooEarlyException
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

/**
 * Translates the `guide` context's exceptions into HTTP responses.
 *
 * Separate from the booking context's handler on purpose: each context owns the mapping of
 * its own vocabulary, and a shared handler would need to import both contexts' exception
 * types — which `ContextRegistryTest` forbids.
 *
 * SDD: see `documentation/use-cases/uc05-start-tour.spec.md` § 3.
 */
@RestControllerAdvice
class GuideExceptionHandler {
    @ExceptionHandler(GuideTourNotFoundException::class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    fun handleGuideTourNotFound(ex: GuideTourNotFoundException) = ErrorResponse(ex.message)

    @ExceptionHandler(InvalidGuideTourStateException::class)
    @ResponseStatus(HttpStatus.CONFLICT)
    fun handleInvalidGuideTourState(ex: InvalidGuideTourStateException) = ErrorResponse(ex.message)

    @ExceptionHandler(TourStartTooEarlyException::class)
    @ResponseStatus(HttpStatus.CONFLICT)
    fun handleTourStartTooEarly(ex: TourStartTooEarlyException) = ErrorResponse(ex.message)

    /** Backstop for identifier parsing on a path variable. See the booking handler. */
    @ExceptionHandler(IllegalArgumentException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleIllegalArgument(ex: IllegalArgumentException) = ErrorResponse(ex.message)

    data class ErrorResponse(
        val error: String?,
    )
}
